Template mode: Infinite loop detected while resolving name references

Hi,

I'm using the template mode and it failed with the error "Infinite loop detected while resolving name references".

- ConfuserEx Version: Confuser.Core 1.6.0-alpha+bb5ce7b9bd, download from [ConfuserEx - AppVeyor ](https://ci.appveyor.com/project/mkaring/confuserex/builds/38100117/artifacts) in https://github.com/mkaring/ConfuserEx/issues/310

- Target Framework:  .NET Framework v4.0

- Operating System: Windows 10

Full log:
```
 [INFO] Confuser.Core 1.6.0-alpha+bb5ce7b9bd Copyright c 2014 Ki, 2018 - 2021 Martin Karing
 [INFO] Running on Microsoft Windows NT 6.2.9200.0, .NET Framework v4.0.30319.42000, 64 bits
[DEBUG] Discovering plugins...
 [INFO] Discovered 13 protections, 1 packers.
[DEBUG] Resolving component dependency...
 [INFO] Loading input modules...
 [INFO] Loading '.\ConfuserExInfiniteLoopBug.dll'...
 [INFO] Initializing...
[DEBUG] Building pipeline...
[DEBUG] Executing 'Type scanner' phase...
 [INFO] Resolving dependencies...
[DEBUG] Checking Strong Name...
[DEBUG] Creating global .cctors...
[DEBUG] Executing 'Name analysis' phase...
[DEBUG] Building VTables & identifier list...
[DEBUG] Analyzing...
 [INFO] Processing module 'ConfuserExInfiniteLoopBug.dll'...
[DEBUG] Executing 'Invalid metadata addition' phase...
[DEBUG] Executing 'Renaming' phase...
[DEBUG] Renaming...
[ERROR] Infinite loop detected while resolving name references.
Processed definition: Original Name: ConfuserExInfiniteLoopBug.inner.ClassA::TestEvent; Name Hash: D82C2349
Assembly: ConfuserExInfiniteLoopBug.dll
Faulty References:
 - Member Override Reference(This Referenced Method(Original Name: ConfuserExInfiniteLoopBug.inner.ClassA::TestEvent; Name Hash: D82C2349); Base Referenced Method(Original Name: ConfuserExInfiniteLoopBug.inner.AEvents::TestEvent; Name Hash: FF497F66))
 - Member Override Reference(This Referenced Method(Original Name: ConfuserExInfiniteLoopBug.inner.ClassA::TestEvent; Name Hash: D82C2349); Base Referenced Method(Original Name: ConfuserExInfiniteLoopBug.inner.BEvents::TestEvent; Name Hash: C182C948))
Failed at 2:27 PM, 0:00 elapsed.
```
I upload a simple demo project to reproduce this:
[ConfuserExInfiniteLoopBug.zip](https://github.com/mkaring/ConfuserEx/files/6341134/ConfuserExInfiniteLoopBug.zip)

Steps to Reproduce:

1. Build the project
2. run _test.bat in cmd

- confuserex.crproj
```
<project outputDir=".\dotfus" baseDir="." xmlns="http://confuser.codeplex.com">
    <rule preset="none" pattern="true" >
        <protection id="anti debug" action="remove" />
        <protection id="anti dump" action="remove" />
        <protection id="anti ildasm" action="remove" />
        <protection id="anti tamper" action="remove" />
        <protection id="constants" action="remove" />
        <protection id="ctrl flow" action="remove" />
        <protection id="invalid metadata" action="remove" />
        <protection id="ref proxy" action="remove" />
        <protection id="rename" action="add">
          <argument name="mode" value="sequential" />
          <argument name="renPublic" value="true" />
          <argument name="flatten" value="false" />
        </protection>
        <protection id="watermark" action="remove" />
        <protection id="resources" action="remove" />
    </rule>
  <module path=".\ConfuserExInfiniteLoopBug.dll" snKey=".\test.snk">
    <rule preset="none" pattern="namespace('ConfuserExInfiniteLoopBug.Public')" >
        <protection id="rename" action="remove"/>
    </rule>
  </module>
  </project>
```

Thanks
Anthony

--- a/Confuser.Renamer/Analyzers/VTableAnalyzer.cs
+++ b/Confuser.Renamer/Analyzers/VTableAnalyzer.cs
@@ -306,7 +306,22 @@ namespace Confuser.Renamer.Analyzers {
 				methodDef.Overrides.Add(new MethodOverride(methodDef, target));
 			}
 			else if (target is IMemberDef targetDef) {
-				CreateOverrideReference(service, methodDef, targetDef);
+				// Reaching this place means that a slot of the base type is overwritten by a specific interface.
+				// In case the this type is implementing the interface responsible for this, we need to declare
+				// this as an override reference. If the base type is implementing the interface (as well), this
+				// declaration is redundant.
+				var overrideRefRequired = true;
+				if (targetDef.DeclaringType.IsInterface) {
+					var baseTypeDef = thisType.BaseType?.ResolveTypeDef();
+					if (!(baseTypeDef is null)) {
+						var baseTypeVTable = service.GetVTables()[baseTypeDef];
+						if (baseTypeVTable.InterfaceSlots.TryGetValue(targetDef.DeclaringType.ToTypeSig(), out var ifcSlots)) {
+							overrideRefRequired = !ifcSlots.Contains(slot);
+						}
+					}
+				}
+				if (overrideRefRequired)
+					CreateOverrideReference(service, methodDef, targetDef);
 			}
 		}
 

--- a/Confuser.Renamer/VTable.cs
+++ b/Confuser.Renamer/VTable.cs
@@ -98,7 +98,7 @@ namespace Confuser.Renamer {
 		internal VTable(TypeSig type) {
 			Type = type;
 			Slots = new List<VTableSlot>();
-			InterfaceSlots = new Dictionary<TypeSig, IList<VTableSlot>>();
+			InterfaceSlots = new Dictionary<TypeSig, IList<VTableSlot>>(TypeEqualityComparer.Instance);
 		}
 
 		public TypeSig Type { get; private set; }
@@ -107,23 +107,11 @@ namespace Confuser.Renamer {
 		public IDictionary<TypeSig, IList<VTableSlot>> InterfaceSlots { get; private set; }
 
 		class VTableConstruction {
-			class TypeSigComparer : IEqualityComparer<TypeSig> {
-				public bool Equals(TypeSig x, TypeSig y) {
-					return new SigComparer().Equals(x, y);
-				}
-
-				public int GetHashCode(TypeSig obj) {
-					return new SigComparer().GetHashCode(obj);
-				}
-
-				public static readonly TypeSigComparer Instance = new TypeSigComparer();
-			}
-
 			// All virtual method slots, excluding interfaces
 			public List<VTableSlot> AllSlots = new List<VTableSlot>();
 			// All visible virtual method slots (i.e. excluded those being shadowed)
 			public Dictionary<VTableSignature, VTableSlot> SlotsMap = new Dictionary<VTableSignature, VTableSlot>();
-			public Dictionary<TypeSig, ILookup<VTableSignature, VTableSlot>> InterfaceSlots = new Dictionary<TypeSig, ILookup<VTableSignature, VTableSlot>>(TypeSigComparer.Instance);
+			public Dictionary<TypeSig, ILookup<VTableSignature, VTableSlot>> InterfaceSlots = new Dictionary<TypeSig, ILookup<VTableSignature, VTableSlot>>(TypeEqualityComparer.Instance);
 		}
 
 		public IEnumerable<VTableSlot> FindSlots(IMethod method) {
@@ -252,7 +240,7 @@ namespace Confuser.Renamer {
 
 			// Populate result V-table
 			ret.InterfaceSlots = vTbl.InterfaceSlots.ToDictionary(
-				kvp => kvp.Key, kvp => (IList<VTableSlot>)kvp.Value.SelectMany(g => g).ToList());
+				kvp => kvp.Key, kvp => (IList<VTableSlot>)kvp.Value.SelectMany(g => g).ToList(), TypeEqualityComparer.Instance);
 
 			foreach (var slot in vTbl.AllSlots) {
 				ret.Slots.Add(slot);

--- a/Confuser2.sln
+++ b/Confuser2.sln
@@ -111,11 +111,15 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "244_ClrProtection", "Tests\
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "244_ClrProtection.Test", "Tests\244_ClrProtection.Test\244_ClrProtection.Test.csproj", "{3ADB8BB1-AE14-49DA-A7E1-1C0D9BEB76E9}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "306_ComplexClassStructureRename.Lib", "Tests\306_ComplexClassStructureRename.Lib\306_ComplexClassStructureRename.Lib.csproj", "{FD93D181-2EC5-4863-8A8F-5F8C84C06B35}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "306_ComplexClassStructureRename.Lib", "Tests\306_ComplexClassStructureRename.Lib\306_ComplexClassStructureRename.Lib.csproj", "{FD93D181-2EC5-4863-8A8F-5F8C84C06B35}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "306_ComplexClassStructureRename", "Tests\306_ComplexClassStructureRename\306_ComplexClassStructureRename.csproj", "{1B52A3D9-014C-4CBF-BB98-09080D9A8D16}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "306_ComplexClassStructureRename", "Tests\306_ComplexClassStructureRename\306_ComplexClassStructureRename.csproj", "{1B52A3D9-014C-4CBF-BB98-09080D9A8D16}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "306_ComplexClassStructureRename.Test", "Tests\306_ComplexClassStructureRename.Test\306_ComplexClassStructureRename.Test.csproj", "{13431429-2DB6-480F-B73F-CA019FE759E3}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "306_ComplexClassStructureRename.Test", "Tests\306_ComplexClassStructureRename.Test\306_ComplexClassStructureRename.Test.csproj", "{13431429-2DB6-480F-B73F-CA019FE759E3}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "342_InterfaceRenamingLoop", "Tests\342_InterfaceRenamingLoop\342_InterfaceRenamingLoop.csproj", "{382B6332-4A57-458D-96EB-B312688A7604}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "342_InterfaceRenamingLoop.Test", "Tests\342_InterfaceRenamingLoop.Test\342_InterfaceRenamingLoop.Test.csproj", "{EC62CE1D-ADD7-419A-84A9-D6A04E866197}"
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
@@ -727,6 +731,30 @@ Global
 		{13431429-2DB6-480F-B73F-CA019FE759E3}.Release|x64.Build.0 = Release|Any CPU
 		{13431429-2DB6-480F-B73F-CA019FE759E3}.Release|x86.ActiveCfg = Release|Any CPU
 		{13431429-2DB6-480F-B73F-CA019FE759E3}.Release|x86.Build.0 = Release|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Debug|x64.Build.0 = Debug|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Debug|x86.ActiveCfg = Debug|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Debug|x86.Build.0 = Debug|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Release|Any CPU.Build.0 = Release|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Release|x64.ActiveCfg = Release|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Release|x64.Build.0 = Release|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Release|x86.ActiveCfg = Release|Any CPU
+		{382B6332-4A57-458D-96EB-B312688A7604}.Release|x86.Build.0 = Release|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Debug|x64.Build.0 = Debug|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Debug|x86.ActiveCfg = Debug|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Debug|x86.Build.0 = Debug|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Release|Any CPU.Build.0 = Release|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Release|x64.ActiveCfg = Release|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Release|x64.Build.0 = Release|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Release|x86.ActiveCfg = Release|Any CPU
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197}.Release|x86.Build.0 = Release|Any CPU
 	EndGlobalSection
 	GlobalSection(SolutionProperties) = preSolution
 		HideSolutionNode = FALSE
@@ -774,6 +802,8 @@ Global
 		{FD93D181-2EC5-4863-8A8F-5F8C84C06B35} = {356BDB31-853E-43BB-8F9A-D8AC08F69EBB}
 		{1B52A3D9-014C-4CBF-BB98-09080D9A8D16} = {356BDB31-853E-43BB-8F9A-D8AC08F69EBB}
 		{13431429-2DB6-480F-B73F-CA019FE759E3} = {356BDB31-853E-43BB-8F9A-D8AC08F69EBB}
+		{382B6332-4A57-458D-96EB-B312688A7604} = {356BDB31-853E-43BB-8F9A-D8AC08F69EBB}
+		{EC62CE1D-ADD7-419A-84A9-D6A04E866197} = {356BDB31-853E-43BB-8F9A-D8AC08F69EBB}
 	EndGlobalSection
 	GlobalSection(ExtensibilityGlobals) = postSolution
 		SolutionGuid = {0D937D9E-E04B-4A68-B639-D4260473A388}

--- a/Tests/342_InterfaceRenamingLoop.Test/342_InterfaceRenamingLoop.Test.csproj
+++ b/Tests/342_InterfaceRenamingLoop.Test/342_InterfaceRenamingLoop.Test.csproj
@@ -1,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net461</TargetFramework>
+    <RootNamespace>InterfaceRenamingLoop.Test</RootNamespace>
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\342_InterfaceRenamingLoop\342_InterfaceRenamingLoop.csproj" />
+    <ProjectReference Include="..\Confuser.UnitTest\Confuser.UnitTest.csproj" />
+  </ItemGroup>
+
+</Project>

--- a/Tests/342_InterfaceRenamingLoop.Test/InterfaceRenamingLoopTest.cs
+++ b/Tests/342_InterfaceRenamingLoop.Test/InterfaceRenamingLoopTest.cs
@@ -1,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Confuser.Core;
+using Confuser.Core.Project;
+using Confuser.Renamer;
+using Confuser.UnitTest;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace InterfaceRenamingLoop.Test {
+	public class InterfaceRenamingLoopTest : TestBase {
+		public InterfaceRenamingLoopTest(ITestOutputHelper outputHelper) : base(outputHelper) { }
+
+		[Theory]
+		[MemberData(nameof(RenameInterfaceLoopTestData))]
+		[Trait("Category", "Protection")]
+		[Trait("Protection", "rename")]
+		[Trait("Issue", "https://github.com/mkaring/ConfuserEx/issues/342")]
+		public async Task RenameInterfaceLoop(string renameMode) =>
+			await Run(
+				"342_InterfaceRenamingLoop.exe",
+				Array.Empty<string>(),
+				new SettingItem<Protection>("rename") {
+					{ "mode", renameMode }
+				},
+				$"_{renameMode}"
+			);
+
+		public static IEnumerable<object[]> RenameInterfaceLoopTestData() {
+			foreach (var renameMode in new[] { nameof(RenameMode.Unicode), nameof(RenameMode.Debug), nameof(RenameMode.Sequential) })
+				yield return new object[] { renameMode };
+		}
+	}
+}

--- a/Tests/342_InterfaceRenamingLoop/342_InterfaceRenamingLoop.csproj
+++ b/Tests/342_InterfaceRenamingLoop/342_InterfaceRenamingLoop.csproj
@@ -1,0 +1,10 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net461</TargetFramework>
+    <RootNamespace>InterfaceRenamingLoop</RootNamespace>
+    <StartupObject>InterfaceRenamingLoop.Program</StartupObject>
+  </PropertyGroup>
+
+</Project>

--- a/Tests/342_InterfaceRenamingLoop/CBase.cs
+++ b/Tests/342_InterfaceRenamingLoop/CBase.cs
@@ -1,0 +1,7 @@
+﻿namespace InterfaceRenamingLoop {
+	internal class CBase {
+		public virtual void TestEvent(int code, string description) {
+
+		}
+	}
+}

--- a/Tests/342_InterfaceRenamingLoop/ClassA.cs
+++ b/Tests/342_InterfaceRenamingLoop/ClassA.cs
@@ -1,0 +1,10 @@
+﻿namespace InterfaceRenamingLoop {
+	internal class ClassA : CBase, IAEvents {
+		public void TestA(int errorCode) {
+
+		}
+		public override void TestEvent(int code, string description) {
+
+		}
+	}
+}

--- a/Tests/342_InterfaceRenamingLoop/ClassB.cs
+++ b/Tests/342_InterfaceRenamingLoop/ClassB.cs
@@ -1,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace InterfaceRenamingLoop {
+	internal class ClassB : ClassA {
+	}
+}

--- a/Tests/342_InterfaceRenamingLoop/IAEvents.cs
+++ b/Tests/342_InterfaceRenamingLoop/IAEvents.cs
@@ -1,0 +1,6 @@
+﻿namespace InterfaceRenamingLoop {
+	internal interface IAEvents {
+		void TestA(int errorCode);
+		void TestEvent(int errorCode, string description);
+	}
+}

--- a/Tests/342_InterfaceRenamingLoop/Program.cs
+++ b/Tests/342_InterfaceRenamingLoop/Program.cs
@@ -1,0 +1,16 @@
+﻿using System;
+
+namespace InterfaceRenamingLoop {
+	public class Program {
+		internal static int Main(string[] args) {
+			Console.WriteLine("START");
+			
+			var test = new ClassB();
+			test.TestEvent(0, "TEST");
+
+			Console.WriteLine("END");
+
+			return 42;
+		}
+	}
+}

